docs/decisions/0006-role-of-xblock.rst

```diff
@@ -0,0 +1,130 @@
+permitted to do in practice. This stabilized the site and brought performance to
+a tolerable (though not great) level, but at a significant cost. Numerous low
+level optimizations and access patterns complicated the code, while at the same
+time compromising some of the expressive power of XBlocks. Many of the API calls
+that *look* dynamic were actually constrainted with unstated assumptions–maximum
+scores are assumed to be consistent across all user, courses are assumed to have
+a Course → Section → Subsection structure, etc.
+
+Reaching beyond these constraints is still possible, but the results are
+unpredictable. The LMS will still render unusual navigational structures if you
+import it directly via OLX, but mobile clients and analytics will likely break.
+You can define different max scores for different students on the same problem,
+but you will likely see bugs in various places where those scores are cached and
+displayed.
+
+Largely because of these issues, the XBlock runtime gradually shifted away from
+the all-encompassing role it once played in courseware:
+
+* Grading and Scheduling became distinct subsystems with their own data models
+  and APIs. Instead of these systems synchronously pulling data from XBlock at
+  query time, we push XBlock data into these systems and query them using their
+  own APIs. This has vastly increased the operational stability of the platform.
+* LTI support has been extracted from ``LtiConsumerXBlock``, so that LTI
+  integration can happen outside the XBlock runtime.
+* There is active planning to add limited QTI support to Open edX.
+* The ``learning_sequence`` application is being developed to take over Section
+  and Sequence level navigation, decoupling those concepts from the XBlock
+  runtime in the LMS.
+* There is a strong interest in creating more adaptive and composable courseware
+  experiences which are difficult to do within the constraints of the XBlock
+  runtime.
+* There have been early conversations around what a successor to XBlock might
+  look like, though no development work has begun.
+
+At the same time, it's important to note that there are hundreds of XBlocks in
+existence and a vast quantity of valuable content that has been written for
+those XBlocks. Preserving the viability of this content is critical to the
+success of Open edX.
+
+
+Decision
+========
+
+XBlocks have provided us a fantastic source of innovative content, at the cost
+of many operational issues. We'll proceed in a way that allows us to preserve
+the most valuable parts of the XBlock framework, while developing a more
+performant, XBlock-agnostic layer of extensible LMS APIs.
+
+In concrete terms:
+
+* XBlocks will continue to be supported at the Unit and individual module level
+  (e.g. ProblemBlock, VideoBlock). XBlocks will execute in a container that only
+  has their Unit (VerticalBlock). They will be allowed to query sibling blocks,
```

Comment:
Is `VerticalBlock` going to be fixed, or flexible to allow other ways of presenting a single unit? Is it worth removing `vertical` from the system as a well, and just giving individual module level blocks ways to access siblings?

Reply:
For what it's worth we also added [`UnitBlock`](https://github.com/edx/edx-platform/blob/60e2b53d4c056c14667ea4b98c631ee127791281/common/lib/xmodule/xmodule/unit_block.py#L18-L28) some time ago, which is like vertical but with no UI nor functionality other than grouping arbitrary blocks together.

Reply:
I think there at least a few facets to `VerticalBlock`'s problems:

**The name `VerticalBlock` is semantically bad.** We should shift it to be something that does not dictate a prefixed layout. That could happen by transforming the data at publish time in some way or other.

**`VerticalBlock` rendering comes with a bunch of cruft and implicit dependencies that aren't needed 95% of the time.** A new XBlock type might be more minimal and a publishing phase scan the contents to be smarter about the dependencies actually needed to render, and store those somewhere so that we're not always including the kitchen sink of JS/CSS from edx-platform. This is where the `UnitBlock` @bradenmacdonald mentions comes in, where it's largely the same idea as a `VerticalBlock`, but lighter.

**`VerticalBlocks` might be pointless altogether, since it's just a flat list of blocks.** As you said, maybe we just kill off `vertical` and add the ability to talk to siblings. Maybe nobody actually even needs siblings at all.

That last question is especially interesting to me because I think there's potentially a fork in the road where the composition of a Unit is something that lives in or outside of the XBlock runtime. So is a Unit an LMS construct and every XBlock in it only sees itself? That's a simpler model, and I think it would cover the majority of use cases we have today.

But at the same time, there are interesting capabilities that we can give if it's more freeform and we allow the XBlock runtime to maintain control of the Unit as a whole and allow XBlocks to nest arbitrarily. Compositional ideas from the early days that we couldn't scale may actually be feasible if they are constrained to the teapot-sized confines of an individual Unit, as opposed to running rampant over large swaths of the course.

I honestly haven't figured out which stance I support between those two options, and I imagine there are other approaches to consider as well. I definitely want to kick XBlock out of high level navigation, and preserve it for individual pieces of leaf-node content like problems, videos, etc. How much the framework gets squeezed out at the Unit level is still pretty fuzzy to me.